airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/AsyncStreamConsumer.java

```diff
@@ -121,18 +120,28 @@ public void accept(final String messageString, final Integer sizeInBytes) throws
    * Deserializes to a {@link PartialAirbyteMessage} which can represent both a Record or a State
    * Message
    *
+   * PartialAirbyteMessage holds either:
+   * <li>entire serialized message string when message is a valid State Message
+   * <li>serialized AirbyteRecordMessage when message is a valid Record Message</li>
+   *
    * @param messageString the string to deserialize
    * @return PartialAirbyteMessage if the message is valid, empty otherwise
    */
-  private Optional<PartialAirbyteMessage> deserializeAirbyteMessage(final String messageString) {
+  @VisibleForTesting
+  public static Optional<PartialAirbyteMessage> deserializeAirbyteMessage(final String messageString) {
+    // TODO: (ryankfu) plumb in the serialized AirbyteStateMessage to match AirbyteRecordMessage and
+    // code parity
     final Optional<PartialAirbyteMessage> messageOptional = Jsons.tryDeserialize(messageString, PartialAirbyteMessage.class)
         .map(partial -> {
-          if (partial.getRecord() != null) {
+          if (partial.getType().equals(Type.RECORD) && partial.getRecord().getData() != null) {
             return partial.withSerialized(partial.getRecord().getData().toString());
-          } else {
+          } else if (partial.getType().equals(Type.STATE)) {
             return partial.withSerialized(messageString);
+          } else {
```

Comment:
@davinchia added another case where I _believe_ we can have bad data which is when messageString is neither a STATE or RECORD message. In this case, this should not be passed to the consumer. I haven't personally seen this happen but this now more closely matches the description that PartialAirbyteMessage should only contain either STATE or RECORD. Lmk if you think this is unnecessary. Not sure if the platform filters out messages

Reply:
We cannot guarantee messages arrive at the Destination ungarbled, so we definitely need to consider this case.

Less decided whether we should fail fast or log an error.. I think we should do the simplest for now and throw a RTE to make sure we don't drop data.

I'm wondering if this happens in practice somehow. We should add logging + test this out before committing this code.

Reply:
Sounds good, I'll add an RTE here, add some logging, publish a new pre-release version, and run it overnight with some resets to see if this ever gets triggered but agree we should fail fast here